### PR TITLE
refactor(edit patient): use react-query when editing a patient

### LINKED ISSUE #2382: create `useUpdatePatient` hook and remove related redux code
## 🚀 Feature Proposal
Create a `useUpdatePatient` hook which will update a patient and remove all related redux code for getting a patient by id. 

## Motivation
- remove redux

## Technical Notes
- start using the `useUpdatePatient` hook for editing a patient
- remove redux code and related tests

### PR DESCRIPTION
**Pull request by giulianovarriale**
Fixes #2382.

**Changes proposed in this pull request:**

- Use react-query for updating a new patient
- Remove redux code that was used for updating a new patient

### CODE REVIEW

**Comment by giulianovarriale** [comment]
The PR is not finished yet, I opened it already for two reasons:
1. I'm participating in hacktoberfest and today is the last day to contribute.
Anyway, I'll finish my contribution because this is the right thing to do and also I'm having fun implementing react-query on the project.
2. I'm wondering what you expect to be tested in here, since the feature remains the same, but the implementation has been changed. For this PR, I was thinking to write a test for the validator and also a test for the hook, what do you think?

File: src/patients/util/validate-patient.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
+import { isAfter, parseISO } from 'date-fns'
+import { isEmpty } from 'lodash'
+import validator from 'validator'
+
+import Patient from '../../shared/model/Patient'
+
+interface IValidationError {
```
Comment (jackcmeyer) [inline src/patients/util/validate-patient.ts:7]: Let's name this as `PatientValidationError`

Comment (jackcmeyer) [reply src/patients/util/validate-patient.ts:7]: It should also extend Error. See other validations on how to do this.

File: src/patients/hooks/useUpdatePatient.tsx [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  const updateError = validatePatient(cleanPatient)
+
+  if (!isEmpty(updateError)) {
+    throw updateError
+  }
+
+  return PatientRepository.saveOrUpdate(cleanPatient)
+}
+
+export default function useUpdatePatient() {
```
Comment (jackcmeyer) [inline src/patients/hooks/useUpdatePatient.tsx:20]: upon success of the mutation, in should update the patient in the cache.

File: src/patients/edit/EditPatient.tsx [hunk 3]
```diff
@@ -74,7 +70,7 @@ const EditPatient = () => {
         patient={patient}
         isEditable
         onChange={onPatientChange}
-        error={updateError}
+        error={updateError ?? undefined}
```
Comment (jackcmeyer) [inline src/patients/edit/EditPatient.tsx:73]: I think we can simply do `updateError`

File: src/patients/util/validate-patient.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  if (patient.emails) {
+    const errors: (string | undefined)[] = []
+    patient.emails.forEach((email) => {
+      if (!validator.isEmail(email.value)) {
+        errors.push('patient.errors.invalidEmail')
+      } else {
+        errors.push(undefined)
+      }
+    })
+    // Only add to error obj if there's an error
```
Comment (jackcmeyer) [inline src/patients/util/validate-patient.ts:69]: I think code is clear enough, we can remove these comments

**Review by jackcmeyer — CHANGES_REQUESTED** [review]

**Comment by jackcmeyer** [comment]
>  For this PR, I was thinking to write a test for the validator and also a test for the hook, what do you think?

I think this sounds perfect.

**Comment by giulianovarriale** [comment]
@jackcmeyer TBH, I didn't write the validator code. I just copied and pasted from the patient-slice file. I think the validator should be out of the scope of this PR. What I'm gonna do is to open a new PR for the validator and once that one is merged we can rebase this one and use it in here. Do you agree?

**Comment by giulianovarriale** [comment]
@jackcmeyer Please, before moving on with this PR, take a look on this one: https://github.com/HospitalRun/hospitalrun-frontend/pull/2471

**Review by tehKapa — DISMISSED** [review]

**Comment by tehKapa** [comment]
@giulianovarriale can you resolve the conflict? thanks

**Comment by giulianovarriale** [comment]
I did that already on my local environment. I'm writing the hook test. Once I finished it, I'll push it.

**Comment by giulianovarriale** [comment]
@tehKapa It is ready for review again.

**Review by tehKapa — APPROVED** [review]

**Pull request closed without merging**